styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,364 @@
+        return candidates.get(new Random().nextInt(candidates.size()));
+      }
+    }
+  }
+
+  public ShardedCounter(Datastore datastore, Time time) {
+    this.datastore = Objects.requireNonNull(datastore);
+    this.time = time;
+  }
+
+  /**
+   * Returns a recent snapshot.
+   */
+  private CounterSnapshot getCounterSnapshot(String counterId) {
+    final CounterSnapshot snapshot = cache.getIfPresent(counterId);
+    if (snapshot != null && snapshot.isRecent()) {
+      return snapshot;
+    }
+    return refreshCounterSnapshot(counterId);
+  }
+
+  /**
+   * Update cached snapshot with most recent state of counter in Datastore
+   */
+  private CounterSnapshot refreshCounterSnapshot(String counterId) {
+    final CounterSnapshot newSnapshot = CounterSnapshot.fromDatastore(datastore, counterId, time);
+    cache.put(counterId, newSnapshot);
+    return newSnapshot;
+  }
+
+  /**
+   * Must be called within a TransactionCallable. (?)
+   *
+   * <p>Augments the transaction with operations to persist the given limit in Datastore. So long as
+   * there has been no preceding successful updateLimit operation, no limit is applied in
+   * updateCounter operations on this counter.
+   */
+  public void updateLimit(DatastoreReaderWriter transaction, String counterId, long limit) {
+    final Key limitKey = datastore.newKeyFactory().setKind(KIND_COUNTER_LIMIT).newKey(counterId);
+
+    transaction.put(Entity.newBuilder(limitKey).set(PROPERTY_LIMIT, limit).build());
+  }
+
+  /**
+   * Reads the latest limit value from Datastore, for the specified {@param counterId}
+   */
+  public static long getLimit(Datastore datastore, String counterId) {
+    final Key limitKey = datastore.newKeyFactory().setKind(KIND_COUNTER_LIMIT).newKey(counterId);
+    final Entity limitEntity = datastore.get(limitKey);
+    if (limitEntity == null) {
+      return Long.MAX_VALUE;
+      // Or IllegalStateException("No limit found in Datastore for " + counterId);?
+    } else {
+      return limitEntity.getLong(PROPERTY_LIMIT);
+    }
+  }
+
+  /**
+   * Must be called within a TransactionCallable. Augments the transaction with certain operations
+   * that strongly consistently increment resp. decrement the counter referred to by counterId, and
+   * cause the transaction to fail to commit if the counter's associated limit is exceeded. Also
+   * spurious failures are possible.
+   *
+   * <p>Delta should be +/-1 for graceful behavior, due to how sharding is currently implemented.
+   * Updates with a larger delta are prone to spuriously fail even when the counter is not near to
+   * exceeding its limit. Failures are certain when delta >= limit / NUM_SHARDS + 1.
+   */
+  public void updateCounter(DatastoreReaderWriter transaction, String counterId, long delta) {
+    CounterSnapshot snapshot = getCounterSnapshot(counterId);
+    int shardIndex;
+    try {
+      shardIndex = snapshot.pickShardWithSpareCapacity(delta);
+    } catch (InvalidShardStateException e) {
```

Comment:
As commented on the throw - it's really not invalid, and catching the condition ought to be unneeded unneeded, it's a rabbit hole to more complexity... The repeated `fetchShards` in the constructor ought to be sufficient for reducing the spurious failures at startup that we didn't like in the stress test, but even that might be unnecessary in continued operation. The algorithm as such is really meant to survive all eventually consistent query behavior. Let's clarify that, if needed? And also test some scenarios where it does so?

Writing code to "handle" the eventual consistency makes that fact ever more hidden - be it two refreshes or a loop of refreshes, we still aren't guaranteed to overcome it, but it's extra code to test (hard to test, too).